src/preprocessing/point_in_poly/hierarchical_winding.jl

```diff
@@ -0,0 +1,217 @@
+                                       child_left, child_right)
+    end
+end
+
+struct HierarchicalWinding{BB}
+    bounding_box::BB
+
+    function HierarchicalWinding(geometry)
+        min_corner = geometry.min_corner .- sqrt(eps())
+        max_corner = geometry.max_corner .+ sqrt(eps())
+
+        directed_edges = zeros(Int, length(geometry.edge_normals))
+
+        bounding_box = BoundingBoxTree(geometry, eachface(geometry), directed_edges,
+                                       min_corner, max_corner)
+
+        return new{typeof(bounding_box)}(bounding_box)
+    end
+end
+
+@inline function (winding::HierarchicalWinding)(mesh, query_point)
+    (; bounding_box) = winding
+
+    return hierarchical_winding(bounding_box, mesh, query_point)
+end
+
+function hierarchical_winding(bounding_box, mesh, query_point)
+    (; min_corner, max_corner) = bounding_box
+
+    if bounding_box.is_leaf
+        return naive_winding(mesh, bounding_box.faces, query_point)
+
+    elseif !is_in_box(query_point, min_corner, max_corner)
+        # `query_point` is outside bounding box
+        return -naive_winding(mesh, bounding_box.closing_faces, query_point)
+    end
+
+    winding_number_left = hierarchical_winding(bounding_box.child_left, mesh, query_point)
+    winding_number_right = hierarchical_winding(bounding_box.child_right, mesh, query_point)
+
+    return winding_number_left + winding_number_right
+end
+
+# This only works when all `vertices` are unique
+function determine_closure!(closing_faces, min_corner, max_corner, mesh::TriangleMesh{3},
+                            faces, directed_edges)
+    (; edge_vertices_ids, face_vertices_ids, face_edges_ids, vertices) = mesh
+
+    directed_edges .= 0
+    intersecting_faces = Int[]
+
+    # Find all exterior edges
+    for face in faces
+        v1 = face_vertices_ids[face][1]
+        v2 = face_vertices_ids[face][2]
+        v3 = face_vertices_ids[face][3]
+
+        if is_in_box(vertices[v1], min_corner, max_corner) &&
+           is_in_box(vertices[v2], min_corner, max_corner) &&
+           is_in_box(vertices[v3], min_corner, max_corner)
+            # Face is completely inside the bounding box
+
+            edge1 = face_edges_ids[face][1]
+            edge2 = face_edges_ids[face][2]
+            edge3 = face_edges_ids[face][3]
+
+            if edge_vertices_ids[edge1] == (v1, v2)
+                directed_edges[edge1] += 1
+            else
+                directed_edges[edge1] -= 1
+            end
+            if edge_vertices_ids[edge2] == (v2, v3)
+                directed_edges[edge2] += 1
+            else
+                directed_edges[edge2] -= 1
+            end
+            if edge_vertices_ids[edge3] == (v3, v1)
+                directed_edges[edge3] += 1
+            else
+                directed_edges[edge3] -= 1
+            end
+
+        else # Face is intersecting the boundaries
+            push!(intersecting_faces, face)
```

Comment:
```suggestion
            # Remove intersecting faces in the calculation of the exterior edges.
            # This way, all exterior edges are inside the bounding box, and so will be the closing surface.
            # The intersecting faces are later added with opposite orientation,
            # so that the closing is actually a closing of the exterior plus intersecting faces.
            push!(intersecting_faces, face)
```